### PR TITLE
Jgfouca/convert case.setup

### PR DESCRIPTION
**Pull request by jgfouca**
NCK_Ld3.f45_g37_rx1.A.melvin_gnu is still failing for me, but I've been on this branch for a while and I'd like to get some feedback/help.

### CODE REVIEW

File: cime_config/acme/machines/taskmaker.pl [hunk 1]
```diff
@@ -81,7 +81,7 @@
 # add the list of ALL *xml fields to a hash
 #-------------------------------------------------------------------------------
 my $parser = XML::LibXML->new( no_blanks => 1);
-my $caseroot = $ENV{CASEROOT};
```
Comment (jgfouca) [inline cime_config/acme/machines/taskmaker.pl]: Don't like passing things by env. My change is a little hacky, but should work fine until taskmaker is ported to python.

File: utils/python/CIME/SystemTests/pem.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def build(self, sharedlib_only=False, model_only=False):
         for bld in range(1,3):
             logging.warn("Starting bld %s"%bld)
             self._case.flush()
-            run_cmd("./case.setup -clean -testmode")
```
Comment (jgfouca) [inline utils/python/CIME/SystemTests/pem.py]: These should be changed to library calls once we get everything working.

File: utils/python/CIME/XML/entry_id.py [hunk 3]
```diff
@@ -106,9 +106,13 @@ def get_value(self, vid, attribute={}, resolved=True, subgroup=None):
         if resolved:
             val = self.get_resolved_value(val)
 
-        # Return value as right type
-        type_str = self._get_type_info(node)
-        return convert_to_type(val, type_str, vid)
+        # Return value as right type if we were able to fully resolve
```
Comment (jgfouca) [inline utils/python/CIME/XML/entry_id.py]: Necessary to handle cross-file dependencies.

File: utils/python/CIME/XML/env_batch.py [hunk 4]
```diff
@@ -24,29 +24,31 @@ def set_value(self, item, value, subgroup=None, ignore_type=False):
...
                 if vnode is not None:
-                    val = EnvBase.set_value(self, vnode, value, ignore_type)
+                    val = EnvBase._set_value(self, vnode, item, value, ignore_type=ignore_type)
 
         return val
 
-    def get_value(self, item, attribute={}, resolved=True, subgroup=None):
+    def get_value(self, item, attribute={}, resolved=True, subgroup="run"):
+        """
+        Must default subgroup to something in order to provide single return value
```
Comment (jgfouca) [inline utils/python/CIME/XML/env_batch.py]: Worth looking at. Need get_value to return a single value.

File: utils/python/CIME/XML/machines.py [hunk 5]
```diff
@@ -244,13 +230,127 @@ def get_module_system_type(self):
...
     def get_module_system_init_path(self, lang):
-        init_nodes = self.get_node("init_path", attributes={"lang":lang})
+        init_nodes = self.get_node("init_path", attributes={"lang":lang}, root=self.machine_node)
         return init_nodes.text
 
     def get_module_system_cmd_path(self, lang):
-        cmd_nodes = self.get_node("cmd_path", attributes={"lang":lang})
+        cmd_nodes = self.get_node("cmd_path", attributes={"lang":lang}, root=self.machine_node)
         return cmd_nodes.text
+
```
Comment (jgfouca) [inline utils/python/CIME/XML/machines.py]: Lotta code needed to handle all the mpi/batch stuff.

File: utils/python/CIME/utils.py [hunk 6]
```diff
@@ -8,16 +8,13 @@
 import os
 import time
 import re
-if sys.version_info[0] == 2:
```
Comment (jgfouca) [inline utils/python/CIME/utils.py]: pylint was flagging this and we only support major version 2 for the forseeable future.

File: cime_config/acme/machines/config_compilers.xml [hunk 7]
```diff
@@ -1120,14 +1116,4 @@ for mct, etc.
 	<ALBANY_PATH>$(ALBANY_PATH)</ALBANY_PATH>
 </compiler>
 
-<compiler MACH="caldera">
```
Comment (jedwards4b) [inline cime_config/acme/machines/config_compilers.xml]: Why was this removed?  This is my machine supporting the acme environment.

File: scripts/Tools/case.setup [hunk 8]
```diff
@@ -1,475 +1,316 @@
...
+        else:
+            logger.debug("Macros script already created ...skipping")
+
+        # Set tasks to 1 if mpi-serial library
+        if case.get_value("MPILIB") == "mpi-serial":
+            for vid, value in case:
+                if vid.startswith("NTASKS_") and value != 1:
+                    case.set_value(vid, 1)
+
+        # Check ninst. JGF: What does NINST stand for? Less hardcoded way to get comps?
```
Comment (jedwards4b) [inline scripts/Tools/case.setup]: In CESM there can be multiple instances of each component model (an ensemble) NINST is the instance of that component.

File: scripts/Tools/case.setup [hunk 9]
```diff
@@ -1,475 +1,316 @@
...
+        for comp in ["ATM", "LND", "ROF", "OCN", "ICE", "GLC", "WAV"]:
```
Comment (jedwards4b) [inline scripts/Tools/case.setup]: The prefered method to get comps is 
        files = Files()
        drv_config_file = files.get_value("CONFIG_DRV_FILE")
        drv_comp = Component(drv_config_file)
        component_classes = drv_comp.get_valid_model_components()

File: scripts/Tools/case.setup [hunk 10]
```diff
@@ -1,475 +1,316 @@
...
+            if os.path.exists("LockedFiles/env_build.xml"):
+                os.remove("LockedFiles/env_build.xml")
+
+            case.flush()
+            check_lockedfiles()
+
+            pestot = int(run_cmd("Tools/taskmaker.pl -sumonly"))
+            case.set_value("TOTALPES", pestot)
+
+            # Compute cost based on PE count
```
Comment (jedwards4b) [inline scripts/Tools/case.setup]: This cost calculation is broken and I think that it should be removed.

File: utils/python/CIME/task_maker.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+class TaskMaker(object):
+
+    def __init__(self, case=None, remove_dead_tasks=False):
+        self.case = case if case is not None else Case()
+        self.remove_dead_tasks = remove_dead_tasks
+
+        self.layout_strings = \
+"""
+COMP_CPL NTASKS_CPL NTHRDS_CPL ROOTPE_CPL PSTRID_CPL
```
Comment (jedwards4b) [inline utils/python/CIME/task_maker.py]: Get rid of hard coded component names as above?

**Comment by jedwards4b** [comment]
Review complete - what's happening with NCK test and how can I help?

File: cime_config/acme/machines/config_compilers.xml (hunk 7 above)
Comment (jgfouca) [inline cime_config/acme/machines/config_compilers.xml]: Ah, sorry. I'll put it back. It was the only block that defined MACH and not COMPILER and I didn't want to support that, but I already had to make change to support it and just forgot to restore this.

File: scripts/Tools/case.setup (hunk 8 above)
Comment (jgfouca) [inline scripts/Tools/case.setup]: Done.

File: scripts/Tools/case.setup (hunk 9 above)
Comment (jgfouca) [inline scripts/Tools/case.setup]: Done.

File: scripts/Tools/case.setup (hunk 10 above)
Comment (jgfouca) [inline scripts/Tools/case.setup]: Want me to just set CCSM_ESTCOST and CCSM_PCOST to zero?

Comment (jedwards4b) [inline scripts/Tools/case.setup]: No, let's just leave it there for now and address it later.

File: utils/python/CIME/task_maker.py (hunk 11 above)
Comment (jgfouca) [inline utils/python/CIME/task_maker.py]: Done.

**Comment by jgfouca** [comment]
@jedwards4b , I've made and pushed all the fixes you suggested. The NCK test from cime_developer is failing during sharedlib build. Could you try it locally and see if you get the same fail?
